src/server/utils/RethinkDataLoader.js

```diff
@@ -0,0 +1,131 @@
+import DataLoader from 'dataloader';
+import getRethink from 'server/database/rethinkDriver';
+import {getUserId} from 'server/utils/authorization';
+
+const defaultCacheKeyFn = (key) => key;
+
+const indexResults = (results, indexField, cacheKeyFn = defaultCacheKeyFn) => {
+  const indexedResults = new Map();
+  results.forEach((res) => {
+    indexedResults.set(cacheKeyFn(res[indexField]), res);
+  });
+  return indexedResults;
+};
+
+const normalizeRethinkDbResults = (keys, indexField, cacheKeyFn = defaultCacheKeyFn) => (results) => {
+  const indexedResults = indexResults(results, indexField, cacheKeyFn);
+  // return keys.map((val) => indexedResults.get(cacheKeyFn(val)));
+  return keys.map((val) => indexedResults.get(cacheKeyFn(val)) || new Error(`Key not found : ${val}`));
+};
+
+const makeStandardLoader = (table) => {
+  const batchFn = async (keys) => {
+    const r = getRethink();
+    const docs = await r.table(table)
+      .getAll(r.args(keys), {index: 'id'});
+    return normalizeRethinkDbResults(keys, 'id')(docs);
+  };
+  return new DataLoader(batchFn);
+};
+
+const makeCustomLoader = (batchFn, options) => {
+  // Make a default {} because if we forget this, then it'll be a tricky memory leak to find DEV only
+  if (!options) {
+    console.warn('Did you forget the options?', batchFn);
+  }
+  return new DataLoader(batchFn, options);
+};
+
+// const groupByField = (arr, fieldName) => {
+//  const obj = {};
+//  for (let ii = 0; ii < arr.length; ii++) {
+//    const doc = arr[ii];
+//    const fieldVal = doc[fieldName];
+//    obj[fieldVal] = obj[fieldVal] || [];
+//    obj[fieldVal].push(doc);
+//  }
+//  return obj;
+// };
+const primeStandardLoader = (loader, items) => {
+  items.forEach((item) => {
+    loader.prime(item.id, item);
+  });
+};
+
+
+export default class RethinkDataLoader {
```

Comment:
yeah, the readme isn't great about realworld examples, however there is one example here: https://github.com/facebook/dataloader#caching-per-request

this is the standard practice in other open source projects, although we all suffer with what to call it! There's a dataloader, then a dataloader bag, then a shared data loader is a bag of dataloader bags.

there's no value in exporting the separately because you don't know what you'll need at the time of creation, unless in misunderstood?